### PR TITLE
Standardise ammo disassembly

### PR DESCRIPTION
**Pull request by SimonPurdie**
Removes the bullet-pulling action from the kinetic bullet puller, and instead use deconstruction/uncrafting recipes to allow ammo to be disassembled like any other item.

To retain existing functionality, a new tag is added for crafting recipes: "uncraft_single_charge"
This allows counted-by-charge items to still be disassembled even if you have less than the default charge amount. The components listed in the recipe are amounts received for disassembling a single charge.
When choosing to disassemble a stack of the item, then either the default charge amount or all remaining charges are disassembled (whichever is smaller). The components returned are multiplied by the number of charges disassembled.
Disassembly always takes the same amount of time, regardless if there a fewer charges than the default.

The old system stricly requires gun skill at 2, and has no chance of component loss. This isn't how general disassembly works. Recipe difficulty is set to 5, which means a high chance of losing components at lower gun skill levels. However this also means you can practice gun skill through ammo disassembly, even at markmanship 0. The difficulty could potentially be set lower/higher to avoid affecting game balance, or adjusted for individual ammmo types.

### CODE REVIEW

File: src/iuse.h [hunk 1]
```diff
@@ -225,13 +224,9 @@ class iuse
     // Helper for handling pesky wannabe-artists
     static int handle_ground_graffiti( player *p, item *it, const std::string prefix );
 
-    static void reset_bullet_pulling();
-    static void load_bullet_pulling(JsonObject &jo);
 protected:
     typedef std::pair<std::string, int> result_t;
     typedef std::vector<result_t> result_list_t;
```
Comment (BevapDin) [inline src/iuse.h]: Are the two typedefs her still used anywhere? If not, they should be removed as well.

File: src/crafting.h [hunk 2]
```diff
@@ -52,6 +52,7 @@ struct recipe {
     std::map<skill_id, int> required_skills;
     bool reversible; // can the item be disassembled?
     bool autolearn; // do we learn it just by leveling skills?
+    bool uncraft_single_charge; // recipe is to uncraft one charge of an item with multiple default charges
```
Comment (BevapDin) [inline src/crafting.h]: This member is not initialized anywhere, reading from it invokes undefined behavior. You need to load it from JSON, probably in a similar way as the other booleans (above) are loaded.

Comment (hitbutton) [inline src/crafting.h]: Big whoops on that, I'm surprised my testing worked. I've fixed the initialisation in crafting.cpp now, and its working as expected for me with uncraft_single_charge set true or false in the json.

**Comment by hitbutton** [comment]
Marking this as WIP, due to PR #14416 possibly introducing recipe flags. That seems like it would be a better way to implement "uncraft_single_charge" uncrafting recipes, instead of bloating the recipe class.

**Comment by hitbutton** [comment]
#14416 is merged now, and I've amended this PR to use a flag instead of adding a class member.

Along with #14724 this should make pulling bullets less like pulling teeth.

**Pull request merged**